lib/iris/tests/unit/coords/test_DimCoord.py

```diff
@@ -0,0 +1,620 @@
+        result = coord.bounds_dtype
+        self.assertTrue(result == test_dtype)
+
+    def test_f16(self):
+        test_dtype = np.float16
+        coord = DimCoord([1], bounds=np.array([[0, 4]], dtype=test_dtype))
+        result = coord.bounds_dtype
+        self.assertTrue(result == test_dtype)
+
+
+def lazyness_string(data):
+    return 'lazy' if isinstance(data, da.Array) else 'real'
+
+
+_DEBUG_ALL_TYPES = _DEBUG
+
+
+def coords_all_dtypes_and_lazynesses(self, dtypes=(np.float64, np.int16)):
+    # Generate coords with all possible types of points and bounds, and all
+    # of the given dtypes.
+    points_types = ['real']
+    bounds_types = ['no', 'real']
+    if _DIMCOORD_LAZY_API:
+        points_types.append('lazy')
+        bounds_types.append('lazy')
+    for dtype in (np.float64, np.int16):
+        for points_type_name in points_types:
+            for bounds_type_name in bounds_types:
+                pts = np.asarray(self.pts_real, dtype=dtype)
+                bds = np.asarray(self.bds_real, dtype=dtype)
+                if points_type_name == 'lazy':
+                    pts = da.from_array(pts, pts.shape)
+                if bounds_type_name == 'lazy':
+                    bds = da.from_array(bds, bds.shape)
+                elif bounds_type_name == 'no':
+                    bds = None
+                coord = DimCoord(pts, bounds=bds)
+                result = (coord, points_type_name, bounds_type_name)
+                if _DEBUG_ALL_TYPES:
+                    msg = '\nTest result type {}({} {})'
+                    print(msg.format(coord.dtype, result[1], result[2]))
+                yield result
+
+
+class Test__getitem__(tests.IrisTest):
+    # Test for DimCoord indexing with various types of points and bounds.
+    def setUp(self):
+        setup_test_arrays(self)
+
+    def test_dtypes(self):
+        # Index coords with all combinations of real+lazy points+bounds, and
+        # either an int or floating dtype.
+        # Check that dtypes are preserved in all cases, taking dtypes directly
+        # from the core points and bounds arrays (as we have no masking).
+        for (main_coord, points_type_name, bounds_type_name) in \
+                coords_all_dtypes_and_lazynesses(self):
+
+            sub_coord = main_coord[:2]
+
+            coord_dtype = main_coord.dtype
+            msg = ('Indexing main_coord of dtype {} '
+                   'with {} points and {} bounds '
+                   'changed dtype of {} to {}.')
+
+            sub_points = _core_points(sub_coord)
+            self.assertEqual(
+                sub_points.dtype, coord_dtype,
+                msg.format(coord_dtype,
+                           points_type_name, bounds_type_name,
+                           'points', sub_points.dtype))
+
+            if bounds_type_name is not 'no':
+                sub_bounds = _core_bounds(sub_coord)
+                self.assertEqual(
+                    sub_bounds.dtype, coord_dtype,
```

Comment:
Yes, good spot, this should be checking against main_coord.bounds_dtype instead.

We do support different types for points + bounds.
I'll have a think about where and how we should test for that ...